File: leetcode/python/Sum Root to Leaf Numbers/Sum_Root_to_Leaf_Numbers.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return an integer
    def sumNumbers(self, root):
        self.res = 0
        num = 0
        self._sumNumbers(root,num)
        return self.res
    
    def _sumNumbers(self,root,num):
        if root is None:
            return
        elif root.left is None and root.right is None:
            num += root.val
            self.res += num
        else:
            num += root.val
            self._sumNumbers(root.left,10*num)
            self._sumNumbers(root.right,10*num)

File: leetcode/cpp/SumRoottoLeafNumbers.cpp
//
// Created by zhangqi on 2018/1/9.
//

//Time : O(n)
//Space : O(h)

#include "TreeNode.h"

class Solution {
public:
    int sumNumbers(TreeNode* root) {
        return sumNumbersRecv(root, 0);
    }

private:
    int sumNumbersRecv(TreeNode *root, int sum){
        if(!root){
            return 0;
        }
        if(!root->left && !root->right){
            return sum * 10 + root->val;
        }

        return sumNumbersRecv(root->left, sum * 10 + root->val ) + sumNumbersRecv(root->right, sum * 10 + root->val) ;
    }
};